src/os/bluestore/BlueStore.cc

```diff
@@ -596,6 +617,121 @@ ostream& operator<<(ostream& out, const BlueStore::Buffer& b)
   return out << ")";
 }
 
+namespace {
+
+/*
+ * Due to a bug in key string encoding (see a comment for append_escaped)
+ * the KeyValueDB iterator does not lexicographically sort the same
+ * way that ghobject_t does: objects with the same hash may have wrong order.
+ *
+ * This is the iterator wrapper that fixes the keys order.
+ */
+
+class CollectionListIterator {
+public:
+  CollectionListIterator(const KeyValueDB::Iterator &it)
+    : m_it(it), m_chunk_iter(m_chunk.begin()) {
+  }
+
+  bool valid() const {
+    return m_key != nullptr;
+  }
+
+  const ghobject_t &oid() const {
+    return *m_oid;
+  }
+
+  const std::string &key() const {
+    return *m_key;
+  }
+
+  void lower_bound(const std::string &key) {
+    ghobject_t oid;
+    int r = get_key_object(key, &oid);
+    ceph_assert(r != -1);
+
+    std::string pkey;
+    _key_encode_prefix(oid, &pkey);
+
+    m_it->lower_bound(pkey);
+    if (!get_next_chunk()) {
+      return;
```

Comment:
Perhaps you need to "invalidate" (i.e. set m_key to nullptr) at this point

Reply:
indeed. Thanks. It looks like better to do this invalidation in `get_next_chunk()` itself.